baselines/fedpara/fedpara/main.py

```diff
@@ -0,0 +1,135 @@
+"""Main script for running FedPara."""
+
+import flwr as fl
+import hydra
+from flwr.common.typing import Config
+from hydra.core.hydra_config import HydraConfig
+from hydra.utils import instantiate
+from omegaconf import DictConfig, OmegaConf
+
+from fedpara import client, server, utils
+from fedpara.dataset import load_datasets
+from fedpara.utils import (
+    get_parameters,
+    save_results_as_pickle,
+    seed_everything,
+    set_client_state_save_path,
+)
+
+
+@hydra.main(config_path="conf", config_name="mnist", version_base=None)
+def main(cfg: DictConfig) -> None:
+    """Run the baseline.
+
+    Parameters
+    ----------
+    cfg : DictConfig
+        An omegaconf object that stores the hydra config.
+    """
+    # 1. Print parsed config
+    print(OmegaConf.to_yaml(cfg))
+    seed_everything(cfg.seed)
+    OmegaConf.to_container(cfg, resolve=True)
+    if "state_path" in cfg:
+        state_path = set_client_state_save_path(cfg.state_path)
+    else:
+        state_path = None
+
+    # 2. Prepare dataset
+    train_loaders, test_loader = load_datasets(
+        config=cfg.dataset_config,
+        num_clients=cfg.num_clients,
+        batch_size=cfg.batch_size,
+    )
+
+    # 3. Define clients
+    # In this scheme the responsability of choosing the client is on the client manager
+
+    client_fn = client.gen_client_fn(
+        train_loaders=train_loaders,
+        test_loader=test_loader,
+        model=cfg.model,
+        num_epochs=cfg.num_epochs,
+        args={"device": cfg.client_device, "algorithm": cfg.algorithm},
+        state_path=state_path,
+    )
+
+    def get_on_fit_config():
+        def fit_config_fn(server_round: int):
+            fit_config: Config = OmegaConf.to_container(  # type: ignore
+                cfg.hyperparams, resolve=True
+            )
+            fit_config["curr_round"] = server_round
+            return fit_config
+
+        return fit_config_fn
+
+    net_glob = instantiate(cfg.model)
+
+    # 4. Define strategy
+    if cfg.strategy.min_evaluate_clients:
+        strategy = instantiate(
+            cfg.strategy,
+            on_fit_config_fn=get_on_fit_config(),
+            initial_parameters=fl.common.ndarrays_to_parameters(
+                get_parameters(net_glob)
+            ),
+        )
+
+    else:
+        evaluate_fn = server.gen_evaluate_fn(
+            num_clients=cfg.num_clients,
+            test_loader=test_loader,
+            model=cfg.model,
+            device=cfg.server_device,
+            state_path=cfg.state_path,
```

Comment:
`state_path` is not present for CIFAR configs so this raises an error. I think the best would be to define `state_path` as:
```
state_path: null
```
for both `cifar10.yaml` and `cifar100.yaml` and then in `get_evaluate_fn()` (and other places if uses) ignore this. But inspecting `get_evaluate_fn` i see it doesn't accept a `state_path` argument.

Reply:
so maybe we should remove that last argument ?